### PR TITLE
tests: Test all client layouts

### PR DESCRIPTION
**Pull request by Elv13**
This is a failing test extracted from the API refactor PR. It fails because of existing problems

For example, `set_master_width_factor` is defined as:

```lua
if mwfact >= 0 and mwfact <= 1 then
```

And a few other corner case (literally) where the clients are pushed to the screen edge with invalid sizes. It prints many warnings, asserts a few time if the titlebars are enables. Xephyr also complain in its own log about invalid operations.

It's rather easy to fix, but the question is where. Who's (object/module) responsibility is it to avoid those errors. Should all layouts check that themselves or should `capi.client` just reject invalid request. A third option is claim we don't care and mute the warnings.

### CODE REVIEW

**Comment by Elv13** [comment]
Rebased. Here is a sample of the (very long) log:

```
2016-04-12 04:09:01 W: awesome: xerror:940: X error: request=ConfigureWindow (major 12, minor 0), error=BadAlloc (11)

2016-04-12 04:09:01 W: awesome: xerror:940: X error: request=ConfigureWindow (major 12, minor 0), error=BadAlloc (11)

2016-04-12 04:09:01 W: awesome: xerror:940: X error: request=CreatePixmap (major 53, minor 0), error=BadAlloc (11)

2016-04-12 04:09:01 W: awesome: xerror:940: X error: request=FreePixmap (major 54, minor 0), error=BadPixmap (4)

2016-04-12 04:09:01 W: awesome: xerror:940: X error: request=ConfigureWindow (major 12, minor 0), error=BadAlloc (11)

2016-04-12 04:09:01 W: awesome: xerror:940: X error: request=ConfigureWindow (major 12, minor 0), error=BadAlloc (11)

2016-04-12 04:09:01 W: awesome: xerror:940: X error: request=CreatePixmap (major 53, minor 0), error=BadAlloc (11)

2016-04-12 04:09:01 W: awesome: xerror:940: X error: request=ConfigureWindow (major 12, minor 0), error=BadAlloc (11)

2016-04-12 04:09:01 W: awesome: xerror:940: X error: request=ConfigureWindow (major 12, minor 0), error=BadAlloc (11)

2016-04-12 04:09:01 W: awesome: xerror:940: X error: request=CreatePixmap (major 53, minor 0), error=BadAlloc (11)

2016-04-12 04:09:01 W: awesome: xerror:940: X error: request=ConfigureWindow (major 12, minor 0), error=BadAlloc (11)

2016-04-12 04:09:01 W: awesome: xerror:940: X error: request=ConfigureWindow (major 12, minor 0), error=BadAlloc (11)

2016-04-12 04:09:01 W: awesome: xerror:940: X error: request=CreatePixmap (major 53, minor 0), error=BadAlloc (11)

2016-04-12 04:09:01 W: awesome: xerror:940: X error: request=CopyArea (major 62, minor 0), error=BadDrawable (9)

2016-04-12 04:09:01 W: awesome: xerror:940: X error: request=CopyArea (major 62, minor 0), error=BadDrawable (9)

2016-04-12 04:09:01 W: awesome: xerror:940: X error: request=CopyArea (major 62, minor 0), error=BadDrawable (9)

2016-04-12 04:09:01 W: awesome: xerror:940: X error: request=CopyArea (major 62, minor 0), error=BadDrawable (9)

2016-04-12 04:09:01 W: awesome: xerror:940: X error: request=FreePixmap (major 54, minor 0), error=BadPixmap (4)

2016-04-12 04:09:01 W: awesome: xerror:940: X error: request=FreePixmap (major 54, minor 0), error=BadPixmap (4)

2016-04-12 04:09:01 W: awesome: xerror:940: X error: request=FreePixmap (major 54, minor 0), error=BadPixmap (4)

2016-04-12 04:09:01 W: awesome: xerror:940: X error: request=FreePixmap (major 54, minor 0), error=BadPixmap (4)

2016-04-12 04:09:02 E: Error during a protected call: lib/wibox/drawable.lua:151: Cairo context entered error state: INVALID_SIZE

stack traceback:

	lib/gears/protected_call.lua:17: in function <lib/gears/protected_call.lua:16>

	[C]: in function 'assert'

	lib/wibox/drawable.lua:151: in function 'do_redraw'

	lib/wibox/drawable.lua:342: in function <lib/wibox/drawable.lua:340>

	(tail call): ?

	[C]: in function 'xpcall'

	lib/gears/protected_call.lua:31: in function <lib/gears/protected_call.lua:29>

	(tail call): ?

	lib/gears/timer.lua:158: in function <lib/gears/timer.lua:156>

2016-04-12 04:09:02 W: awesome: xerror:940: X error: request=ConfigureWindow (major 12, minor 0), error=BadAlloc (11)

2016-04-12 04:09:02 W: awesome: xerror:940: X error: request=ConfigureWindow (major 12, minor 0), error=BadAlloc (11)

2016-04-12 04:09:02 W: awesome: xerror:940: X error: request=CreatePixmap (major 53, minor 0), error=BadAlloc (11)

2016-04-12 04:09:02 W: awesome: xerror:940: X error: request=CopyArea (major 62, minor 0), error=BadDrawable (9)

2016-04-12 04:09:02 W: awesome: xerror:940: X error: request=FreePixmap (major 54, minor 0), error=BadPixmap (4)

2016-04-12 04:09:15 E: Error during a protected call: lib/wibox/drawable.lua:151: Cairo context entered error state: INVALID_SIZE

stack traceback:

	lib/gears/protected_call.lua:17: in function <lib/gears/protected_call.lua:16>

	[C]: in function 'assert'

	lib/wibox/drawable.lua:151: in function 'do_redraw'

	lib/wibox/drawable.lua:342: in function <lib/wibox/drawable.lua:340>

	(tail call): ?

	[C]: in function 'xpcall'

	lib/gears/protected_call.lua:31: in function <lib/gears/protected_call.lua:29>

	(tail call): ?

	lib/gears/timer.lua:158: in function <lib/gears/timer.lua:156>
```

**Comment by psychon** [comment]
Looking at this via xtrace (and accidentally freezing my computer in the process), here are some of the reasons for the X11 errors:

Some window (frame window + the actual terminal(?)) is resized to a huge size. The X11 server fails to allocate the necessary memory (BadAlloc error); when we later try to draw its titlebar, another BadAlloc occurs:
```
29.989 000:<:16f3: 28: Request(12): ConfigureWindow window=0x00600071 values={x=10 y=30 width=65530 height=726}
29.989 000:<:16f4: 28: Request(12): ConfigureWindow window=0x00a0000d values={x=0 y=20 width=65530 height=706}
```
Note that 65530 is suspiciously close to 2^16 (65536) and thus there is likely some overflow or truncation involved here.
The following patch
```diff
diff --git a/objects/client.c b/objects/client.c
index d530376..400bf82 100644
--- a/objects/client.c
+++ b/objects/client.c
@@ -2579,6 +2579,9 @@ luaA_client_geometry(lua_State *L)
             geometry.height = luaA_getopt_number(L, 2, "height", c->geometry.height);
         }
 
+        if (geometry.width >= 6000)
+            luaL_dostring(L, "print(debug.traceback(\"Huge resize\"))");
+
         client_resize(c, geometry, c->size_hints_honor);
     }
```
gives the following backtrace for the first occurrence of this:
```
Huge resize
stack traceback:
	[string "print(debug.traceback("Huge resize"))"]:1: in main chunk
	[C]: in function 'geometry'
	...hon/projects/awesome/build/lib/awful/layout/init.lua:179: in function <...hon/projects/awesome/build/lib/awful/layout/init.lua:164>
	(tail call): ?
	[C]: in function 'xpcall'
	.../projects/awesome/build/lib/gears/protected_call.lua:31: in function <.../projects/awesome/build/lib/gears/protected_call.lua:29>
	(tail call): ?
	...e/psychon/projects/awesome/build/lib/gears/timer.lua:158: in function <...e/psychon/projects/awesome/build/lib/gears/timer.lua:156>
```
Hmm... not really useful.

**Comment by psychon** [comment]
I added `require("gears.debug").dump({something, g})` to `awful.layout.arrange` right before its call to `c:geometry(g)`:
```
  2 : table: 0x163eca0
    height : 187 (number)
    x : 1024 (number)
    y : 200 (number)
    width : -2 (number)
```
The layout assigns a size that is smaller than the value of `useless_gap` to a client. After the `useless_gap`-subtraction, the value becomes negative. The C code does not check this in any way (it only catches value zero) and `-2` becomes `2^16-2` (yay for integer truncation).

**Comment by Elv13** [comment]
Thanks for looking into this, so the question I asked above remain. Is it the C core job to prevent that or it is the layout job? (also, what about negative border_width, border_width larger than the screen and height smaller than the titlebar).

**Comment by psychon** [comment]
Oh, sorry. I will write a patch that makes the C code ignore resizes to negative sizes, just like it already ignores size 0. Something will also have to be done about the layouts, I guess... perhaps `awful.layout.arrange` should enforce a minimum size of 0?
Yup, `border_width` has the same issue with negative sizes. For "high positive sizes", I'd just blame the user/the theme. 
Yup, client height smaller than the titlebar's height is another problem. :-( I'll look into it.

File: luaa.h [hunk 1]
```diff
@@ -153,6 +153,33 @@ luaA_getopt_number(lua_State *L, int idx, const char *name, lua_Number def)
...
+}
+
+static inline int
+luaA_getopt_number_range(lua_State *L, int idx, const char *name, lua_Number def, double min, double max)
+{
+    lua_getfield(L, idx, name);
+    if (lua_isnil(L, -1) || lua_isnumber(L, -1))
+        def = luaA_optnumber_range(L, -1, def, min, max);
+    lua_pop(L, 1);
+    return def;
```
Comment (psychon) [inline luaa.h]: Could you turn all of the `double`s here into `lua_Number`? I don't really have a good reason for this, but the rest of the file uses `lua_Number` as well.

**Comment by psychon** [comment]
Looks fine to me. Thanks for this. One minor comment above (which I don't really see as important, feel free to ignore).
